sway-core/src/ir_generation/compile.rs

```diff
@@ -155,16 +156,30 @@ pub(super) fn compile_function(
         let args = ast_fn_decl
             .parameters
             .iter()
-            .map(|param| {
-                convert_resolved_typeid(context, &param.type_id, &param.type_span)
-                    .map(|ty| (param.name.as_str().into(), ty, param.name.span()))
-            })
+            .map(|param| convert_fn_param(context, param))
             .collect::<Result<Vec<(String, Type, Span)>, CompileError>>()?;
 
         compile_fn_with_args(context, md_mgr, module, ast_fn_decl, args, None).map(&Some)
     }
 }
 
+fn convert_fn_param(
+    context: &mut Context,
+    param: &TypedFunctionParameter,
+) -> Result<(String, Type, Span), CompileError> {
+    convert_resolved_typeid(context, &param.type_id, &param.type_span).map(|ty| {
+        Ok((
+            param.name.as_str().into(),
+            if param.is_mutable && look_up_type_id(param.type_id).is_copy_type() {
+                Type::Pointer(Pointer::new(context, ty, param.is_mutable, None))
+            } else {
+                ty
+            },
+            param.name.span(),
+        ))
+    })?
```

Comment:
I think you can remove that `Ok` and the `?` on line 180?

`convert_resolved_typeid()` returns a `Result`, and so will the call to `.map()`, so there's no need to introduce another `Result`.